src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/FileWatchers/SolutionConfigFileWatcher.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using NuGet.Configuration;
+
+namespace NuGet.PackageManagement.VisualStudio.Utility.FileWatchers
+{
+    internal class SolutionConfigFileWatcher : IFileWatcher
```

Comment:
I'm slightly concerned about the fact that we have to reimplement the logic for config reading. 

Can we just watch the files we loaded instead? 

I understand that won't handle new file creation, but maybe that's a worthwhile trade off?

Reply:
I understand the concern. If in some point in the future we change where config files are loaded from, then the VS file watchers need to be updated as well, otherwise there'll be a bug. 

When I worked on this implementation, I made the tradeoff that removing paths from being loaded would be a breaking change, and therefore incredibly unlikely. Also, I'm only aware of a single time when a new path was added (`UserSettingsPath\config\*.config`).  Basically, it's a kind of "public API", and we try hard not to change public APIs much. This allowed me to keep the implementation simple, while still "powerful" (like checking for file creation, and minimizing FileWatcher new instances & disposal).

I'm also wondering if customers creating new nuget.config files in the solution directory is actually uncommon. When I need to test a customer scenario, particularly for non-SDK projects, I create a solution in VS, open the terminal to run `dotnet new nugetconfig`, then immediate reload the solution so it gets loaded.  I can imagine customers who are starting new solutions also benefit from the solution directory file creation watching.

If you feel particularly strongly, please let me know. Otherwise I'd like to hear from other people about their thoughts.

Reply:
I guess an alternative would be to only watch the solution folder for new files. 

@NuGet/nuget-client 
Any thoughts on this discussion?

If no one else shares my concerns, we can keep the current iteration.

Reply:
> I guess an alternative would be to only watch the solution folder for new files.

+1 for this suggestion. Looking at the comments in https://github.com/NuGet/Home/issues/1538, I get the impression that that customers are mainly looking for `NuGet.Config file under the solution root`. For example, [`branch switch, git pull scenario`](https://github.com/NuGet/Home/issues/1538#issuecomment-425571925) Andrew pointed out in the issue. Chances of customers modifying repository level NuGet.Config file is higher than modifying the hierarchy of NuGet.Config files.

In the [docs](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file), we have made clear that `Add a nuget.config file in the root of your project repository. This is considered a best practice as it promotes repeatability and ensures that different users have the same NuGet configuration. You may need to configure clear elements to ensure no user or machine specific configuration is applied.`

Reply:
Ok, I've changed it.

Keep in mind that if a customer has two instances of VS open, and they use VS to change package sources, which gets persisted in the user-config directory (rather than a solution nuget.config), this means that the second instance of VS is not going to automatically get the changes, and the customer will still need to reload the solution.

Also, I can't remember where I saw this directory layout, but I've seen at least one repo that puts the solution file in the repo `src\` directory, but the `nuget.config` is still in the repo root. When branch switching that repo, this PR will no longer detect the `nuget.config` file change.

Reply:
> Also, I can't remember where I saw this directory layout, but I've seen at least one repo that puts the solution file in the repo src\ directory, but the nuget.config is still in the repo root. When branch switching that repo, this PR will no longer detect the nuget.config file change.

This makes me change my mind about it :D 

Is there a perf hit from watching multiple directories? If not, then our concerns are moot. 

I'd love for this to be a topic we don't have to revisit :)

Reply:
> Is there a perf hit from watching multiple directories?

The only perf issue related to file system watchers I'm aware of are:

1. being too slow to process events, causing the watcher's queue to overflow, and then events are lost
2. multiple events coming in causing the event handler to run multiple times (need to "debounce" the event).

However, in my opinion neither is a concern for us, because:
1. all we do is set up a new lazy, which is fairly fast. Honestly, it might be just as fast as adding a message to a concurrent queue, so I'm not sure we could meaningfully make the handler faster.
2. our handler is idempotent
3. the probability of our watch filter triggering a large number of events is low
4. regardless of which filenames trigger the event, the action is always the same. We don't do action 1 for some files and action 2 for different files.

I know that on Linux (not sure about Mac), there is a limit to the number of file system watchers allowed (fs.inotify.max_user_watches), but I've never heard of such a case with Windows, nor can I quickly find any evidence of any limits on Windows. And the code in this PR only runs on Windows. A long time ago, Windows did have a low max of total handles that could be opened. I don't know if that limit was significantly increased to the point of irrelevance when Windows became 32-bit or 64-bit, or maybe the limit was independent of the OS's bit-ness. But since VS2022 is 64-bit only, I'm confident we don't have to worry about handle count.

So, all this to say that I'm not aware of any perf issues with many file system watchers that rarely trigger any events.